lib/ts/controllers/s-modal.ts

```diff
@@ -0,0 +1,222 @@
+            var returnElementSelector = this.data.get("return-element");
+            if (returnElementSelector) {
+                this.returnElement = <HTMLElement>document.querySelector(returnElementSelector);
+
+                if (!this.returnElement) {
+                    throw "Unable to find element by return-element selector: " + returnElementSelector;
+                }
+            }
+        }
+
+        /**
+         * Toggles the visibility of the modal element
+         * @param show Optional parameter that force shows/hides the element or toggles it if left undefined
+         */
+        _toggle (show?: boolean | undefined) {
+            var toShow = show;
+
+            // if we're letting the class toggle, we need to figure out if the popover is visible manually
+            if (typeof toShow === "undefined") {
+                toShow = this.modalTarget.getAttribute("aria-hidden") === "true";
+            }
+
+            // show/hide events trigger before toggling the class
+            this.triggerEvent(toShow ? "show" : "hide");
+
+            this.modalTarget.setAttribute("aria-hidden", toShow ? "false" : "true");
+
+            if (toShow) {
+                this._bindDocumentEvents();
+            }
+            else {
+                this._unbindDocumentEvents();
+                this._focusReturnElement();
+            }
+
+            // check for transitionend support
+            var supportsTransitionEnd = (<HTMLElement>this.modalTarget).ontransitionend !== undefined;
+
+            // shown/hidden events trigger after toggling the class
+            if (supportsTransitionEnd) {
+                // wait until after the modal finishes transitioning to fire the event
+                this.modalTarget.addEventListener("transitionend", () => {
+                    //TODO this is firing waaay to soon?
+                    this.triggerEvent(toShow ? "shown" : "hidden");
+                }, { once: true });
+            } else {
+                this.triggerEvent(toShow ? "shown" : "hidden");
+            }           
+        }
+
+        /**
+         * Listens for the s-modal:hidden event and focuses the returnElement when it is fired
+         */
+        _focusReturnElement() {
+            if (!this.returnElement) {
+                return;
+            }
+
+            this.modalTarget.addEventListener("s-modal:hidden", () => {
+                // double check the element still exists when the event is called
+                if (this.returnElement && document.body.contains(this.returnElement)) {
+                    this.returnElement.focus();
+                }
+            });
+        }
+
+        /**
+         * Binds tab presses on tabbable items such that tabbing only works within the modal
+         */
+        _bindTabFocusTrap() {
+            // get all tabbable items
+            var allTabbables = Array.from(this.modalTarget.querySelectorAll("[href], input, select, textarea, button, [tabindex]"))
```

Comment:
If we're doing this right, we should move `allTabbables` et al inside `_boundTabTrap`.  That way we capture elements being added and removed.

It also means we don't need a whole `_bindTabFocusTrap` method.

Reply:
We still need the `_bindFocusTrap` method so I can unbind it from the document on close. I agree on adding `allTabbables` into the method so we can capture changes, but I'm going to punt this for now. I'm going to revisit s-modal very soon for a second round of changes and this'll be one of 'em.